Feature/speech input complete silence length

rouddy opened the pull request:
add speechInputPossiblyCompleteSilenceLengthMs option and modify result

sowens-csd reviewed (commented):
Thanks for the PR, looks like it was a fair amount of work. 

Have you tried this on Android and does it affect the behaviour? For example if you set a 20 second value will the device wait 20 seconds before terminating the listen session? Is there a limit to how long a delay you can set? 

My previous attempts to use various options on Android didn't change the recognizers behaviour at all. This looks like it might be a different option than the ones I've tried though, I don't remember.

rouddy commented:
i cannot find the documents about limit. but when i tried 60000ms on android, it works. and there was no side effect on ios. i didn't tried on macos but it might. but i just noticed that i don't change the code for web. i'll do additional commit maybe today or tommorow. and i changed some test code, but there should be descent of coverige.

rouddy commented:
i modified web code and now it work on web when i tested..

library is not supported on windows according to documentation but there is some codes for windows.
so i changed windows code too.

rouddy commented:
when i make speech input complete silence length longer on android, android returns final result without complete listening.
so i have to distinguish final result without complete and final result with complete. so i named  final result without complete as intermediate.

sowens-csd reviewed (commented):
So I see you added a Timer. I think that is to detect that speech recognition is final even though Android doesn't indicate that it is? 

Given that timer I wonder if we still need the new intermediate state? That new state adds complexity for every platform and I'm not sure it provides enough value on Android to make it worth the complexity. What do you think?  (I could still be not understanding the value, so feel free to educate me.)

rouddy commented:
on android onEndOfSpeech is called when user pause speaking. so onEndOfSpeech sometimes is called when speech recognition is not over even if i do not change any codes. because onEndOfSpeech method is not the callback for recognition ends but the callback for user stop speaking. you can see [report](https://github.com/csdcorp/speech_to_text/issues/622) i wrote.

and i setup the timer starts when onEndOfSpeech not onBeginningOfSpeech. which means notifyListening(false) is already called if i don't modify. i just postpone the completion waiting for another speech starting with timer.

intermediate state makes complexity but we should distinguish final result with recognition complete and final result without recognition complete.
on android recognition returns final result without complete, and buffer is cleared when final result is returned.

try on android with set speechInputPossiblyCompleteSilenceLengthMs over 5000. you can see buffer is cleared when pause speaking without recognition completion.

if i can do refactoring whole codes, i would make stream for show finalization state to decrease complexity.

sowens-csd reviewed (approved):
Thanks for the explanation, very helpful. I'm merging this, any further changes we can discuss as a new issue. I'll release a beta with these changes and we can hopefully get some feedback from other users. 

This has been a frequently requested improvement so I think you'll make a number of Android users very happy with this improvement.
